resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java

```diff
@@ -1021,17 +1021,16 @@ public <T extends Throwable> ExceptionMapper<T> getExceptionMapperForClass(Class
     //   @Override
     public <T> AsyncResponseProvider<T> getAsyncResponseProvider(Class<T> type) {
         Class asyncType = type;
-        AsyncResponseProvider<T> mapper = null;
-        while (mapper == null) {
-            if (asyncType == null)
-                break;
-            Map<Class<?>, AsyncResponseProvider> asyncResponseProviders = getAsyncResponseProviders();
-            if (asyncResponseProviders != null)
-                mapper = asyncResponseProviders.get(asyncType);
-            if (mapper == null)
-                asyncType = asyncType.getSuperclass();
+
+        final Map<Class<?>, AsyncResponseProvider> asyncResponseProviders = getAsyncResponseProviders();
+        if (asyncResponseProviders == null)
+            return null;
+        for (Class<?> aClass : asyncResponseProviders.keySet()) {
+            if (aClass.isAssignableFrom(asyncType)) {
+                return asyncResponseProviders.get(aClass);
+            }
```

Comment:
I'm not sure we should remove the `type.getSuperClass()` checking. That's a change in behavior. My assumption is this is done so we don't check the `CompletionStatge`, but maybe I'm wrong.

Reply:
@jamezp sorry for delay, I must have missed the notification

I added back the type.getSuperClass check. I assumed it was an attempt to check if the return type extends  `CompletionStage` but maybe there's another case that it was meant to handle